autoray/autoray.py

```diff
@@ -419,28 +409,75 @@ def complex_add_re_im(re, im):
     return re + 1j * im
 
 
+# ----------------------------- Custom dispatchers -------------------------- #
+
+
+def register_dispatch(fun, dispatcher):
+    _DISPATCHERS[fun] = dispatcher
+
+
+def default_dispatcher(*args, **kwargs):
+    return args[0]
+
+
+# lookup of custom dispatcher methods, for cases when backend cannot be inferred
+#     accurately from first argument.
+_DISPATCHERS = defaultdict(lambda: default_dispatcher)
+
+
+# Numpy array joining operations always take a sequence of arrays as input.
+#     The backend is inferred from the first element of this sequence. Numpy
+#     always assumes that the argument is iterbale, so calling args[0][0]
+#     should be safe.
+def join_array_dispatcher(*args, **kwargs):
+    # NOTE: To be super safe, we could do a try / except, or
+    #       hasattr(args[0], '__getitem__'). hasattr is then probably better,
+    #       since there would be multiple types of errors to catch
+    return args[0][0]
+
+
+# List of functions listed in numpy API as array joining operations
+register_dispatch("concatenate", join_array_dispatcher)
+register_dispatch("stack", join_array_dispatcher)
+register_dispatch("block", join_array_dispatcher)
+register_dispatch("vstack", join_array_dispatcher)
+register_dispatch("hstack", join_array_dispatcher)
+register_dispatch("dstack", join_array_dispatcher)
+register_dispatch("column_stack", join_array_dispatcher)
+register_dispatch("row_stack", join_array_dispatcher)
+
+
+# einsum takes string as first argument a string, but backend should
+#      be inferred from second argument.
+def einsum_dispatcher(*args, **kwargs):
+    if isinstance(args[0], str):
+        return args[1]
+    return args[0]

```

Comment:
Just to elaborate here, `einsum` has the two following formats:
```python
einsum('ab,bc->ca', x, y)
einsum(x, [0, 1], y, [1, 2], [2, 0])
```
thus line 455 which is uncovered by tests currently. I think fine as is, just thought I'd mention it.

Reply:
Ah, right I forgot about the second format. I think it's worth updating the tests to see if this works as intended for all libraries.